pkg/pillar/types/global.go

```diff
@@ -103,312 +107,417 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+)
 
-type PerAgentSettings struct {
-	LogLevel       string // What we log to files
-	RemoteLogLevel string // What we log to zedcloud
-}
+type AgentSettingKey string
+
+const (
+	LogLevel       AgentSettingKey = "loglevel"
+	RemoteLogLevel AgentSettingKey = "remote.loglevel"
+)
+
+type ConfigItemType uint8
+
+const (
+	CONFIGITEM_TYPE_INT ConfigItemType = iota + 1
+	CONFIGITEM_TYPE_BOOL
+	CONFIGITEM_TYPE_STRING
+	CONFIGITEM_TYPE_TRISTATE
+)
+
+type ConfigItemSpec struct {
+	Key      string
+	ItemType ConfigItemType
+
+	IntMin     uint32
+	IntMax     uint32
+	IntDefault uint32
 
-// Default values until/unless we receive them from the cloud
-// We do a GET of config every 60 seconds,
-// PUT of metrics every 60 seconds,
-// If we don't hear anything from the cloud in a week, then we reboot,
-// and during a post-update boot that time is reduced to 10 minutes.
-// On reboot if we can't get a config, then we use a saved one if the saved is
-// not older than 10 minutes.
-// A downloaded image which isn't used is garbage collected after 10 minutes.
-// If a instance has been removed its read/write vdisks are deleted after
-// one hour.
-var GlobalConfigDefaults = GlobalConfig{
-	ConfigInterval:          60,
-	MetricInterval:          60,
-	ResetIfCloudGoneTime:    7 * 24 * 3600,
-	FallbackIfCloudGoneTime: 300,
-	MintimeUpdateSuccess:    600,
-
-	NetworkGeoRedoTime:        3600, // 1 hour
-	NetworkGeoRetryTime:       600,  // 10 minutes
-	NetworkTestDuration:       30,
-	NetworkTestInterval:       300, // 5 minutes
-	NetworkTestBetterInterval: 0,   // Disabled
-	NetworkFallbackAnyEth:     TS_ENABLED,
-	NetworkTestTimeout:        15,
-
-	NetworkSendTimeout: 120,
-
-	UsbAccess:           true, // Contoller likely to default to false
-	SshAccess:           true, // Contoller likely to default to false
-	SshAuthorizedKeys:   "",
-	StaleConfigTime:     600,  // Use stale config for up to 10 minutes
-	DownloadGCTime:      600,  // 10 minutes
-	VdiskGCTime:         3600, // 1 hour
-	DownloadRetryTime:   600,  // 10 minutes
-	DomainBootRetryTime: 600,  // 10 minutes
-	RktGCGracePeriod:    3600, // 1 hour
-
-	AllowNonFreeAppImages:  TS_ENABLED,
-	AllowNonFreeBaseImages: TS_ENABLED,
-
-	DefaultLogLevel:       "info", // XXX Should we change to warning?
-	DefaultRemoteLogLevel: "info", // XXX Should we change to warning?
-
-	Dom0MinDiskUsagePercent: 20,
-	IgnoreDiskCheckForApps:  false,
+	StringDefault   string
+	BoolDefault     bool
+	TriStateDefault TriState
 }
 
-// Check which values are set and which should come from defaults
-// Zero integers means to use default
-func ApplyGlobalConfig(newgc GlobalConfig) GlobalConfig {
+type ConfigItemSpecMap struct {
+	// GlobalSettings key is the agent name.
+	GlobalSettings map[string]ConfigItemSpec
+	// AgentSettings key is the agent name. The inner map's key is the per-agent setting name.
+	AgentSettings map[string]map[string]ConfigItemSpec
```

Comment:
Hmm.. Do we need 2-level for AgentSettings here for SpecMap also? Currently, we only support settings that are applicable to all agents - But can be set differently for each one.. I don't think we need a 2-level map here..

AgentSettings map[AgentSettingKey]ConfigItemSpec